OrbitLinuxTracing/TracerThread.cpp

```diff
@@ -62,71 +62,89 @@ void TracerThread::Run(
       std::move(uprobes_unwinding_visitor));
 
   if (trace_instrumented_functions_) {
+    absl::flat_hash_map<int32_t, int> uprobes_ring_buffer_fds_per_cpu;
+
     for (const auto& function : instrumented_functions_) {
-      std::vector<int> function_uprobes_fds;
-      std::vector<int> function_uretprobes_fds;
-      std::vector<PerfEventRingBuffer> function_uprobes_ring_buffers;
+      absl::flat_hash_map<int32_t, int> function_uprobes_fds_per_cpu;
+      absl::flat_hash_map<int32_t, int> function_uretprobes_fds_per_cpu;
       bool function_uprobes_open_error = false;
 
       for (int32_t cpu : cpuset_cpus) {
         int uprobes_fd = uprobes_stack_event_open(
             function.BinaryPath().c_str(), function.FileOffset(), -1, cpu);
-        std::string buffer_name = absl::StrFormat(
-            "uprobe_retprobe_%#016lx_%u", function.VirtualAddress(), cpu);
-        PerfEventRingBuffer uprobes_ring_buffer{
-            uprobes_fd, BIG_RING_BUFFER_SIZE_KB, buffer_name};
-        if (uprobes_ring_buffer.IsOpen()) {
-          function_uprobes_fds.push_back(uprobes_fd);
-          function_uprobes_ring_buffers.push_back(
-              std::move(uprobes_ring_buffer));
-        } else {
+        if (uprobes_fd < 0) {
           function_uprobes_open_error = true;
-          perf_event_open_errors = true;
-          uprobes_event_open_errors = true;
           break;
         }
 
         int uretprobes_fd = uretprobes_event_open(
             function.BinaryPath().c_str(), function.FileOffset(), -1, cpu);
-        if (uretprobes_fd >= 0) {
-          function_uretprobes_fds.push_back(uretprobes_fd);
-        } else {
+        if (uretprobes_fd < 0) {
```

Comment:
This is leaking uprobes_fd, either scope or close it here...

Reply:
Thank you! I moved `function_uprobes_fds_per_cpu.emplace(cpu, uprobes_fd);` so that uprobes_fd is closed with the other fds to close in case of an error.